lib/logstash/outputs/tcp.rb

```diff
@@ -75,23 +75,32 @@ def register
       @logger.info("Starting tcp output listener", :address => "#{@host}:#{@port}")
       @server_socket = TCPServer.new(@host, @port)
       @client_threads = []
+      @client_threads_lock = Mutex.new
 
       @accept_thread = Thread.new(@server_socket) do |server_socket|
         loop do
-          client_thread = Thread.start(server_socket.accept) do |client_socket|
-            client = Client.new(client_socket, @logger)
-            Thread.current[:client] = client
-            client.run
+          @client_threads_lock.synchronize do
+            @client_threads << Thread.start(server_socket.accept) do |client_socket|
+              begin
+                client = Client.new(client_socket, @logger)
+                Thread.current[:client] = client
+                client.run
+              ensure
+                @client_threads_lock.synchronize do
+                  @client_threads.delete(Thread.current)
+                end
+              end
+            end
           end
-          @client_threads << client_thread
```

Comment:
instead of doing a "long" lock, I'd suggest keeping the `client_thread` var usage and simply do a short lock to update the `@client_threads` ivar like that
```ruby
@client_threads_lock.synchronize do
  @client_threads << client_thread
end
```